### ISSUE TITLE
Support for Map DataType

### ISSUE THREAD

**Issue by chitralverma** [pola-rs/polars#8385]
### Problem description

Most data processing systems/ data frame libs have a non-strict MapType (dict/ HashMap), are any plans to support this in Polars (rust/ py) as well?

Ref arrow type:
https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html#variant.Map
https://arrow.apache.org/docs/python/generated/pyarrow.map_.html#pyarrow.map_

**Comment by ritchie46** [comment]
I don't think it is worth the extra code bloat and complexity. It is a `List<struct<2>>` physically, so that's how we read them in polars.

I think `FixedSizeList` and `Decimal` have much, much higher prio.

**Comment by chitralverma** [comment]
> I don't think it is worth the extra code bloat and complexity. It is a `List<struct<2>>` physically, so that's how we read them in polars.
> 
> I think `FixedSizeList` and `Decimal` have much, much higher prio.

Sure. just added a tracking ticket for now or later.

**Comment by john-sungjin** [comment]
I was wondering if there were any plans to implement maps, now that fixedsizelist and decimal have been added. I have some dicts with variable keys that I'd like to work with; is there a simple workaround with the List<struct<2>> type?

**Comment by theelderbeever** [comment]
> I don't think it is worth the extra code bloat and complexity. It is a `List<struct<2>>` physically, so that's how we read them in polars.
> 
> I think `FixedSizeList` and `Decimal` have much, much higher prio.

@ritchie46 Would the `Map`/`List<struct<2>>` type solve issues like #10234? I run into the empty struct issue quite regularly when dealing with api response data (which `polars` has been a game changer for dealing with) however, if any empty dictionaries are in the response payload `polars` outright fails to parse the data into a dataframe. And when responses a deeply nested it can be pretty untenable to attempt to remove all empty dictionaries.

**Comment by OneCyrus** [comment]
we have issues with duckdb generating an arrow maptype which is not directly supported. it would be great to have polars to completely support the maptype for compatibility with arrow.

**Comment by rodrigogonegit** [comment]
Any news on this?

**Comment by JankoJerinic** [comment]
Can we have an update on this issue?

**Comment by JankoJerinic** [comment]
Specifically - why is this an issue with polars 1.21.0 but not with, say, 1.14.0? Same data loads perfectly fine with an older polars version.

**Comment by JankoJerinic** [comment]
Actually - this is an interesting one. The issue only surfaces when the input PyArrow column is empty, otherwise things work out just fine. Try running this on polars 1.21.0.

```
import pyarrow as pa
import polars as pl

map_data = []
map_array = pa.array(map_data, type=pa.map_(pa.string(), pa.string()))
pa_table = pa.table([map_array], names=['generic_map'])
df = pl.from_arrow(pa_table)
```

This will fail with a fairly misleading message:
```
thread '<unnamed>' panicked at crates/polars-core/src/datatypes/field.rs:234:19:
Arrow datatype Map(Field { name: "entries", dtype: Struct([Field { name: "key", dtype: Utf8, is_nullable: false, metadata: None }, Field { name: "value", dtype: Utf8, is_nullable: true, metadata: None }]), is_nullable: false, metadata: None }, false) not supported by Polars. You probably need to activate that data-type feature.
```

Now, put something in the ``map_data``, say:
```
map_data = [
    {'foo': 'bar'},
    {'foo': 'bar', 'baz': 'qoo'},
    {'foo': 'bar', 'baz': 'qoo', 'qoo': 'qux'},
]
```

Polars will happily create the Series, with the correct type, as seen below. So, this is strictly an issue when the input table has no rows. 

```
List(Struct({'key': String, 'value': String}))
shape: (3, 1)
┌─────────────────────────────────┐
│ generic_map                     │
│ ---                             │
│ list[struct[2]]                 │
╞═════════════════════════════════╡
│ [{"foo","bar"}]                 │
│ [{"foo","bar"}, {"baz","qoo"}]  │
│ [{"foo","bar"}, {"baz","qoo"},… │
└─────────────────────────────────┘
```

**Comment by JankoJerinic** [comment]
> [@JankoJerinic](https://github.com/JankoJerinic) You would need to provide more details.
> 
> Ideally by opening a new issue with a reproducible example.

Also - thank you for the quick reply and apologies for my initial terseness. I realized in real time that this actually only impacts conversion of empty tables.

**Comment by JankoJerinic** [comment]
I just confirmed it, given a ``column_type`` which is a PyArrow ``MapType``, you can "trick" Polars into correctly converting an empty input table, without error, if you "recode" the input column into a list of structs.
```
if isinstance(column_type, pa.MapType) and len(table) == 0:
    recoded_column = pa.array(
                [],
                type=pa.list_(pa.struct([("key", column_type.key_type), ("value", column_type.item_type)])),
            )
```

Resulting Polars data frame:
```
List(Struct({'key': String, 'value': String}))
shape: (0, 1)
┌─────────────────┐
│ generic_map     │
│ ---             │
│ list[struct[2]] │
╞═════════════════╡
└─────────────────┘

```

**Comment by erikbrinkman** [comment]
Is this something where a pr would be welcomed? It seems like longer term it'd be significant work but it could start off small. Something like:

- [ ] add DataType (is this breaking?)
- [ ] Add array and mutable array variants
- [ ] Add serde for io where it would otherwise fail (is this breaking?)
- [ ] add conversion to/from struct
- [ ] Add serde for "native" map types that are currently converted to structs (behind a config)
- [ ] Add other necessary helpers
    - [ ] `concat_map`
    - [ ] `keys`, `values`
    - [ ] `eval_keys`, `eval_values`
    - [ ] `__get_item__`
    - [ ] `len`
    - [ ] others...

where'd it'd provide some value after it can serde files it couldn't otherwise.

**Comment by danielgafni** [comment]
I am reading `Map(K,V)` data from ClickHouse and would really appreciate to have it as `Map` in Polars. 

I have a generic CLI which targets different DBs and storage formats, and Polars is the only part of the system which doesn't support `Map`. For exampole: ClickHouse, DuckDB, PyArrow, and Ibis all support it.  This makes the issue really annoying to me, and I am ready to contribute as well. 

@erikbrinkman are you still interested in this feature? Perhaps we can start as a Polars plugin (custom datatype)?

**Comment by erikbrinkman** [comment]
@danielgafni I'm still semi-intersted in. I've worked around it by just using `List(Struct(key: , value:))` and it feels like the extension type would just be that with a little sugar on top since that's also how arrow treats map types. Since it's just a small amount of sugar, I'm struggling to see how much of a benefit it is to add a whole plugin when writing expression wrappers around the former mostly suffices, and it feels like a lot of overhead if it's not just builtin.

**Comment by danielgafni** [comment]
Alright, in my case there is no work around this issue: 
1. One of the promises of my [project](https://docs.metaxy.io/latest/) is the ability to swap databases and dataframe backends (Ibis, Polars, etc) without changing user code. I cannot treat Polars different to other dataframe backends because of this reason. I cannot control which dataframe flavor does the user of my project decide to use. 
2. Attempting to write `List(Struct(key, value))` to something like Parquet or DeltaLake (or other stroage formats) would result in a wrong datatype in storage, which would make it incompatible with other dataframe backends which expect to work with a `Map`.  

It seems like I will have to move my project to use DuckDB at user API boundaries, and abandon `pl.LazyFrame` support until this issue is resolved. Hopefully sooner than later! 

Meanwhile I started to look into implementing `Map` for Polars myself, and given that physically Arrow's `Map(K,V)` is indeed just `List(Struct(K,V))` it should not be that hard to achieve this. Please let me know if you'll be willing to take a look at some point :)

**Comment by coastalwhite** [comment]
Maybe a much simpler solution is to support arrow export of the arrow `Map` type through the `arrow_schema` argument.

Polars already supports reading `ArrowDataType::Map` in Parquet (and Feather maybe?) by converting to `List(Struct { .. })`, so if `List(Struct { .. })` could be exported to `Map` that would make sense. Not a core team member, so I don't have a say, but I could see that being a suitable solution.

**Comment by danielgafni** [comment]
It would help me. Not ideal and won't solve all the problems, but at least storage schemas would be consistent.

**Comment by erikbrinkman** [comment]
@danielgafni not sure if this actually suits your needs, but I went ahead and created [`polars-map`](https://pypi.org/project/polars-map/). It uses unstable ExtensionTypes to wrap the `List(Struct(...))` storage format. Extension types are a little rough around the edges, but this should provide the metadata you need for serialization, and the methods by default preserve uniqueness.

In order to avoid complexity, this is all done in python, so some of the methods aren't efficient, and because ExtensionTypes are a little rough, many of the functions require the gil to wrap the type at execution time. That said, it should all hopefully work.

**Comment by danielgafni** [comment]
Wow @erikbrinkman, thanks a lot!

Yes, I think it covers my needs, I will test it later this week. I'll probably open a few PRs if something goes wrong.

Not sure how would it work with writing to `DeltaLake` tho... probably should work given it points at `map_` as the associated Arrow type.  

P.S. I already tried the `ExtensionType` route myself, but was unable to get the desired UX from it. It looks like I was missing the custom `Expr` class... well done!

**Comment by danielgafni** [comment]
Thank you @erikbrinkman, I'm supper happy now! The **Map Hell** has been solved for me! I was able to get the behavior I needed for my project (performance at this particular is not a priority as well). I also opened a PR to `polars-map` which improves type checking for `.map` namespace.  

I am also [planning to work](https://github.com/narwhals-dev/narwhals/issues/3525) on `narwhals-map` to bring `Map` support there (as a plugin for now).  

---

That being said, it would be great to see `Map` datatype on Polars' backlog. I believe it's a core type in the Arrow ecosystem and it should be supported (at least for feature parity with :duck: )

**Comment by danielgafni** [comment]
Hey, what's the current stance of Polars maintainers on the `Map` datatype issue? 

I'm asking for two reasons: 
1. It seems like the sooner this problem is solved the better it is for Polars, because it would unfortunately be a breaking change for pipelines that read datasets containing `Map` types. Users will have to rewrite their code that currently expects`List<Struct(...)>` instead of `Map`. 
2. A more personal reason: looks like having first-class `Map` support is too important in my project and I might have to (but don't want to!) switch from Polars to DuckDB as the local compute backend that does support it. I would very much prefer to stay on Polars instead given there is a reasonable timeline for `Map` support. 

I believe support may be opt-in (via `polars.Config`?) at first but should eventually become the default (for correctness reasons).

**Comment by ritchie46** [comment]
After #28894 we discussed internally. It is time for us to support this `Map` types.

**Comment by chitralverma** [comment]
@ritchie46 is there a pr for this or WIP, would be happy to take this up

**Comment by danielgafni** [comment]
@chitralverma I had a draft https://github.com/pola-rs/polars/pull/27000, maybe it will be a bit helpful
